webapp/src/ts/services/rules-engine.service.ts

```diff
@@ -210,27 +204,57 @@ export class RulesEngineService implements OnDestroy {
     };
   }
 
-  private monitorChanges(rulesEngineContext) {
+  private monitorContactChanges() {
     const isReport = doc => doc.type === 'data_record' && !!doc.form;
 
     const dirtyContactsSubscription = this.changesService.subscribe({
-      key: 'mark-contacts-dirty',
+      key: this.CHANGE_WATCHER_KEY,
       filter: change => !!change.doc && (this.contactTypesService.includes(change.doc) || isReport(change.doc)),
       callback: change => {
-        const subjectIds = isReport(change.doc) ? RegistrationUtils.getSubjectId(change.doc) : change.id;
-        const trackPerformance = this.performanceService.track();
+        const subjectIds = [isReport(change.doc) ? RegistrationUtils.getSubjectId(change.doc) : change.id];
+
+        if (this.debounceActive[this.CHANGE_WATCHER_KEY]?.active) {
+          const oldSubjectIds = this.debounceActive[this.CHANGE_WATCHER_KEY].params;
+          if (oldSubjectIds) {
+            subjectIds.push(...oldSubjectIds);
+          }
 
-        return this.rulesEngineCore
-          .updateEmissionsFor(subjectIds)
-          .then((result) => {
-            this.observable.next(subjectIds);
-            trackPerformance?.stop({ name: [ 'rules-engine', 'update-emissions' ].join(':') });
-            return result;
-          });
+          this.debounceActive[this.CHANGE_WATCHER_KEY].params = subjectIds;
+          this.debounceActive[this.CHANGE_WATCHER_KEY].debounceRef();
+          return;
+        }
+
+        const trackPerformance = this.performanceService.track();
+        let resolve;
+        const promise = new Promise(r => resolve = r);
+
+        const debounceRef = _debounce(async () => {
+          this.debounceActive[this.CHANGE_WATCHER_KEY].active = false;
+          this.debounceActive[this.CHANGE_WATCHER_KEY].resolve();
+
+          await this.rulesEngineCore.updateEmissionsFor(this.debounceActive[this.CHANGE_WATCHER_KEY].params);
+          this.observable.next(this.debounceActive[this.CHANGE_WATCHER_KEY].params);
+          trackPerformance?.stop({ name: [ 'rules-engine', 'update-emissions' ].join(':') });
+        }, 1000);
+
+        this.debounceActive[this.CHANGE_WATCHER_KEY] = {
+          active: true,
+          promise: promise,
+          resolve: resolve,
+          performance: {
+            name: [ 'rules-engine', 'update-emissions', 'cancel' ],
+            track: this.performanceService.track()
+          },
+          debounceRef: debounceRef,
+          params: subjectIds,
+        };
+        this.debounceActive[this.CHANGE_WATCHER_KEY].debounceRef();
```

Comment:
I think this callback handler is complicated enough that it deserves to be its own named function...

Reply:
Agree!